### ISSUE TITLE
Qualys WAS - Unicode Styrings with Encoding declaration are not supported

### ISSUE THREAD

**Issue by benweber2** [HASecuritySolutions/VulnWhisperer#156]
**Describe the bug**
When attempting to connect to Qualys for WAS data, an error is shown:
```
INFO:main:Running vulnwhisperer for section qualys_web
INFO:vulnWhispererBase:Connected to database at /opt/VulnWhisperer/data/database/report_tracker.db
INFO:qualysapi.util:Finished building connector.
INFO:qualysWhisperAPI:Connected to Qualys at qualysapi.qualys.com
QualysGuard Username:
QualysGuard Password:
ERROR:main:Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
ERROR: Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
```
**Affected module**
Qualys WAS

**To Reproduce**
Follow setup guys and attempt first download of WAS data

**Expected behavior**
ETL of qualys WAS data.

**System in which VulnWhisperer runs (please complete the following information):**
 - OS: "Ubuntu 18.04.2 LTS"
 - Version 1.7.1

**Comment by reactor7171** [comment]
Hi @benweber2, 

Could you make sure you are using qualysapi version 5.1.0? If you do `pip freeze`, it will give you the list of python libraries installed with its version, and qualysapi 5.1.0 (the latest) is the one that supports VulnWhisperer.

If it is not, please do `pip uninstall qualysapi` and install it again. This new version is a very recent release, we were maintaining our own fork of qualysapi until now.

Cheers!

**Comment by benweber2** [comment]
Hey @qmontal,

I've confirmed I'm using qualysapi 5.1.0:

user@elk:~$ python -V
Python 2.7.15rc1
user@elk:~$ pip freeze | grep qualys
qualysapi==5.1.0

Thanks!

**Comment by reactor7171** [comment]
Could you put the debug trail of the execution? (`-d` flag when executing)
Everytime I had this error was due to a wrong qualysapi version being installed, so if it is not the case, we will need to troubleshoot it.

In the meantime, you can also try uninstalling the qualysapi version from pip, and install the fork that the master branch has running the following command from VulnWhisperer's root folder:
`git submodule init && git submodule update && cd deps/qualysapi && python setup.pi install`

Lets see if that version of qualysapi still gives the same result, otherwise, it might be an issue on the PR that we did to the official qualysapi library.

**Comment by reactor7171** [comment]
For more insight on this issue, the part of the code that breaks is the `qualysapi` module when trying to login, that it should be getting the username and password variables from the config file, but when it doesn't get the expected variables, it changes to interactive mode requesting them; at that moment, as it is a script, it breaks because new stuff that supposedly should be executed encounters an input request.

**Comment by benweber2** [comment]
Debug Output (By the looks of it, it's completing successfully but not finding any scans?):
```
INFO:main:Running vulnwhisperer for section qualys_web
DEBUG:vwConfig:Calling get for qualys_web:enabled
DEBUG:vwConfig:Calling get for qualys_web:hostname
DEBUG:vwConfig:Calling get for qualys_web:username
DEBUG:vwConfig:Calling get for qualys_web:password
DEBUG:vwConfig:Calling get for qualys_web:write_path
DEBUG:vwConfig:Calling get for qualys_web:db_path
DEBUG:vwConfig:Calling getbool for qualys_web:verbose
INFO:vulnWhispererBase:Connected to database at /opt/VulnWhisperer/data/database/report_tracker.db
DEBUG:qualysapi.config:[('max_retries', '10'), ('hostname', 'qualysapi.qualys.com'), ('template_id', '126024'), ('enabled', 'true'), ('username', 'REDACTED'), ('password', 'REDACTED'), ('write_path', '/opt/VulnWhisperer/data/qualys/'), ('db_path', '/opt/VulnWhisperer/data/database'), ('verbose', 'true')]
DEBUG:qualysapi.connector:proxies =
None
DEBUG:qualysapi.connector:max_retries =
10
DEBUG:urllib3.util.retry:Converted retries value: 10 -> Retry(total=10, connect=None, read=None, redirect=None, status=None)
DEBUG:urllib3.util.retry:Converted retries value: 10 -> Retry(total=10, connect=None, read=None, redirect=None, status=None)
INFO:qualysapi.util:Finished building connector.
INFO:qualysWhisperAPI:Connected to Qualys at qualysapi.qualys.com
QualysGuard Username: raymn_sa1
QualysGuard Password:
DEBUG:qualysapi.config:[('max_retries', '3'), ('hostname', 'qualysapi.qualys.com'), ('template_id', '0'), ('username', 'REDACTED'), ('password', 'REDACTED')]
DEBUG:qualysapi.connector:api_call =
/count/was/wasscan
DEBUG:qualysapi.connector:api_version =
None
DEBUG:qualysapi.connector:data <type 'lxml.etree._Element'> =
 <Element ServiceRequest at 0x7fe2f7350ea8>
DEBUG:qualysapi.connector:http_method =
None
DEBUG:qualysapi.connector:concurrent_scans_retries =
0
DEBUG:qualysapi.connector:concurrent_scans_retry_delay =
0
DEBUG:qualysapi.connector:api_call post strip =
count/was/wasscan
DEBUG:qualysapi.connector:Base url =
https://qualysapi.qualys.com/qps/rest/3.0/
DEBUG:qualysapi.connector:headers =
{'X-Requested-With': 'Parag Baxi QualysAPI (python) v5.1.0'}
DEBUG:qualysapi.connector:http_method =
post
DEBUG:qualysapi.connector:api_call post strip =
count/was/wasscan
DEBUG:qualysapi.connector:api_call =
count/was/wasscan
DEBUG:qualysapi.connector:Converting lxml.builder.E to string
DEBUG:qualysapi.connector:Converted:
<ServiceRequest><filters><Criteria operator="EQUALS" field="status">FINISHED</Criteria></filters></ServiceRequest>
DEBUG:qualysapi.connector:url =
https://qualysapi.qualys.com/qps/rest/3.0/count/was/wasscan
DEBUG:qualysapi.connector:data =
<ServiceRequest><filters><Criteria operator="EQUALS" field="status">FINISHED</Criteria></filters></ServiceRequest>
DEBUG:qualysapi.connector:headers =
{'Content-type': 'text/xml', 'X-Requested-With': 'Parag Baxi QualysAPI (python) v5.1.0'}
DEBUG:qualysapi.connector:POST request.
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): qualysapi.qualys.com:443
DEBUG:urllib3.connectionpool:https://qualysapi.qualys.com:443 "POST /qps/rest/3.0/count/was/wasscan HTTP/1.1" 200 None
DEBUG:qualysapi.connector:response headers =
{'Strict-Transport-Security': 'max-age=31536000', 'X-Content-Type-Options': 'nosniff', 'Content-Encoding': 'gzip', 'Transfer-Encoding': 'chunked', 'evaluation': 'false', 'X-Powered-By': 'Qualys:USPOD01:9c0e25e2-ccbc-5af6-e040-10ac13043f6a:9bfbc8d4-ae54-7614-8325-84ddf170ae87', 'Vary': 'Accept-Encoding', 'Server': 'Apache', 'Date': 'Tue, 19 Mar 2019 16:10:05 GMT', 'X-Frame-Options': 'SAMEORIGIN', 'Content-Type': 'application/xml'}
DEBUG:qualysapi.connector:'x-ratelimit-remaining'
DEBUG:qualysapi.connector:response text =
<?xml version="1.0" encoding="UTF-8"?>
<ServiceResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://qualysapi.qualys.com/qps/xsd/3.0/was/wasscan.xsd">
  <responseCode>SUCCESS</responseCode>
  <count>0</count>
</ServiceResponse>
ERROR:main:Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
ERROR: Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
```

FWIW, it still went into interactive mode asking for UN and PW.

**Comment by benweber2** [comment]
Tried removing qualysapi:

`pip uninstall qualysapi`

Confirming library no longer exists:

`pip freeze | grep qualys`

Ran your above script - Confirmed new Qualys installed:

`pip freeze | grep qualys
qualysapi==5.0.4`

Output is similar

`<ServiceResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://qualysapi.qualys.com/qps/xsd/3.0/was/wasscan.xsd">
  <responseCode>SUCCESS</responseCode>
  <count>0</count>
</ServiceResponse>
ERROR:main:Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
ERROR: Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.`

I'm starting to think this is an issue with my permissions on my account, not with Vuln Whisperer.

**Comment by benweber2** [comment]
Confirmed to be an issue on my end.

To others, make sure your API account has access to others scans, not just your own.

**Comment by reactor7171** [comment]
Thanks for the confirmation!

**Comment by benweber2** [comment]
Hey @qmontal - I apologize for the indecisiveness in my closing/reopening - I'm still having this error (moved from Ubuntu18 to Ubuntu16 for better logstash support.)

I'm successfully getting back a scan count now, but still the same Unicode error:

`DEBUG:qualysapi.connector:response headers =
{'Strict-Transport-Security': 'max-age=31536000', 'X-Content-Type-Options': 'nosniff', 'Content-Encoding': 'gzip', 'Transfer-Encoding': 'chunked', 'evaluation': 'false', 'X-Powered-By': 'Qualys:USPOD01:9c0e25e2-ccbc-5af6-e040-10ac13043f6a:9bfbc8d4-ae54-7614-8325-84ddf170ae87', 'Vary': 'Accept-Encoding', 'Server': 'Apache', 'Date': 'Tue, 19 Mar 2019 19:55:54 GMT', 'X-Frame-Options': 'SAMEORIGIN', 'Content-Type': 'application/xml'}
DEBUG:qualysapi.connector:'x-ratelimit-remaining'
DEBUG:qualysapi.connector:response text =
<?xml version="1.0" encoding="UTF-8"?>
<ServiceResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://qualysapi.qualys.com/qps/xsd/3.0/was/wasscan.xsd">
  <responseCode>SUCCESS</responseCode>
  <count>16263</count>
</ServiceResponse>
ERROR:main:Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.
ERROR: Unicode strings with encoding declaration are not supported. Please use bytes input or XML fragments without declaration.`

**Comment by reactor7171** [comment]
If you get into slack, tomorrow we can try to check the issue together, as it will be easier to troubleshoot the issue and make sure if there is some code not working as expected and fix it ^^

I don't have access to a Qualys WAS instance myself, so unfortunately I haven't been able to test and debug that module.

**Comment by reactor7171** [comment]
Hey @benweber2,

I did a couple of commits on the beta-1.8 branch, one should solve the interactive credentials issue, and the other the encoding issue error (that is as far as I could test without a WAS instance). 

If this is the case, I hope this was the only place were the encoding is being an issue on the Qualys WAS module, otherwise I might need to put some more `encode('utf-8')` on the module; this was something I had seen in the Qualys VM side, but I thought qualysapi now was taking care of the encoding issues.

Tell me more whenever you are able to test the latest version, remember to first do `git checkout beta-1.8` in order to get the latest commits.

Cheers!

**Comment by benweber2** [comment]
Hey @qmontal,

Thanks a bunch for your time on this.  Unfortunately I don't see much of a difference in output.  I did notice that it looks like the config is being set twice, once correctly (with the correct template ID), and once without:

`DEBUG:qualysapi.config:[('max_retries', '10'), ('hostname', 'qualysapi.qualys.com'), ('template_id', '126024'), ('enabled', 'true'), ('username', 'redacted'), ('password', 'redacted'), ('write_path', '/opt/VulnWhisperer/data/qualys/'), ('db_path', '/opt/VulnWhisperer/data/database'), ('verbose', 'true')]`

`DEBUG:qualysapi.config:[('max_retries', '3'), ('hostname', 'qualysapi.qualys.com'), ('template_id', '0'), ('username', 'redacted'), ('password', 'redacted')]`

I did still get an interactive prompt for Qualys though.

Current Branch:
`user1@ubuntu:/app/VulnWhisperer$ git branch
* beta-1.8
  master
`
Current Qualysapi library version:
`qualysapi==5.1.0`

Full Debug Logs:
```
INFO:main:Running vulnwhisperer for section qualys_web
DEBUG:vwConfig:Calling get for qualys_web:enabled
DEBUG:vwConfig:Calling get for qualys_web:hostname
DEBUG:vwConfig:Calling get for qualys_web:username
DEBUG:vwConfig:Calling get for qualys_web:password
DEBUG:vwConfig:Calling get for qualys_web:write_path
DEBUG:vwConfig:Calling get for qualys_web:db_path
DEBUG:vwConfig:Calling getbool for qualys_web:verbose
INFO:vulnWhispererBase:Connected to database at /opt/VulnWhisperer/data/database/report_tracker.db
DEBUG:qualysapi.config:[('max_retries', '10'), ('hostname', 'qualysapi.qualys.com'), ('template_id', '126024'), ('enabled', 'true'), ('username', 'redacted'), ('password', 'redacted'), ('write_path', '/opt/VulnWhisperer/data/qualys/'), ('db_path', '/opt/VulnWhisperer/data/database'), ('verbose', 'true')]
DEBUG:qualysapi.connector:proxies =
None
DEBUG:qualysapi.connector:max_retries =
10
DEBUG:urllib3.util.retry:Converted retries value: 10 -> Retry(total=10, connect=None, read=None, redirect=None, status=None)
DEBUG:urllib3.util.retry:Converted retries value: 10 -> Retry(total=10, connect=None, read=None, redirect=None, status=None)
INFO:qualysapi.util:Finished building connector.
INFO:qualysWhisperAPI:Connected to Qualys at qualysapi.qualys.com
QualysGuard Username: redacted
QualysGuard Password:
DEBUG:qualysapi.config:[('max_retries', '3'), ('hostname', 'qualysapi.qualys.com'), ('template_id', '0'), ('username', 'redacted'), ('password', 'redacted')]
DEBUG:qualysapi.connector:api_call =
/count/was/wasscan
DEBUG:qualysapi.connector:api_version =
None
DEBUG:qualysapi.connector:data <type 'lxml.etree._Element'> =
 <Element ServiceRequest at 0x7f993c465518>
DEBUG:qualysapi.connector:http_method =
None
DEBUG:qualysapi.connector:concurrent_scans_retries =
0
DEBUG:qualysapi.connector:concurrent_scans_retry_delay =
0
DEBUG:qualysapi.connector:api_call post strip =
count/was/wasscan
DEBUG:qualysapi.connector:Base url =
https://qualysapi.qualys.com/qps/rest/3.0/
DEBUG:qualysapi.connector:headers =
{'X-Requested-With': 'Parag Baxi QualysAPI (python) v5.1.0'}
DEBUG:qualysapi.connector:http_method =
post
DEBUG:qualysapi.connector:api_call post strip =
count/was/wasscan
DEBUG:qualysapi.connector:api_call =
count/was/wasscan
DEBUG:qualysapi.connector:Converting lxml.builder.E to string
DEBUG:qualysapi.connector:Converted:
<ServiceRequest><filters><Criteria operator="EQUALS" field="status">FINISHED</Criteria></filters></ServiceRequest>
DEBUG:qualysapi.connector:url =
https://qualysapi.qualys.com/qps/rest/3.0/count/was/wasscan
DEBUG:qualysapi.connector:data =
<ServiceRequest><filters><Criteria operator="EQUALS" field="status">FINISHED</Criteria></filters></ServiceRequest>
DEBUG:qualysapi.connector:headers =
{'Content-type': 'text/xml', 'X-Requested-With': 'Parag Baxi QualysAPI (python) v5.1.0'}
DEBUG:qualysapi.connector:POST request.
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): qualysapi.qualys.com
DEBUG:urllib3.connectionpool:https://qualysapi.qualys.com:443 "POST /qps/rest/3.0/count/was/wasscan HTTP/1.1" 200 None
DEBUG:qualysapi.connector:response headers =
{'Strict-Transport-Security': 'max-age=31536000', 'X-Content-Type-Options': 'nosniff', 'Content-Encoding': 'gzip', 'Transfer-Encoding': 'chunked', 'evaluation': 'false', 'X-Powered-By': 'Qualys:USPOD01:9c0e25e2-ccbc-5af6-e040-10ac13043f6a:9bfbc8d4-ae54-7614-8325-84ddf170ae87', 'Vary': 'Accept-Encoding', 'Server': 'Apache', 'Date': 'Wed, 20 Mar 2019 11:02:31 GMT', 'X-Frame-Options': 'SAMEORIGIN', 'Content-Type': 'application/xml'}
DEBUG:qualysapi.connector:'x-ratelimit-remaining'
DEBUG:qualysapi.connector:response text =
<?xml version="1.0" encoding="UTF-8"?>
<ServiceResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://qualysapi.qualys.com/qps/xsd/3.0/was/wasscan.xsd">
  <responseCode>SUCCESS</responseCode>
  <count>16278</count>
</ServiceResponse>
```

**Comment by reactor7171** [comment]
Hi @benweber2,
Can you make sure you pulled the latest results from the beta-1.8 branch? It looks like you changed to beta-1.8 branch from the already downloaded code without the latest changes. In my screen this looks different:

```
2019-03-20 12:33:19 quim main[24241] INFO Running vulnwhisperer for section qualys_web
2019-03-20 12:33:19 quim vwConfig[24241] DEBUG Calling get for qualys_web:enabled
2019-03-20 12:33:19 quim vwConfig[24241] DEBUG Calling get for qualys_web:hostname
2019-03-20 12:33:19 quim vwConfig[24241] DEBUG Calling get for qualys_web:username
2019-03-20 12:33:19 quim vwConfig[24241] DEBUG Calling get for qualys_web:password
2019-03-20 12:33:19 quim vwConfig[24241] DEBUG Calling get for qualys_web:write_path
2019-03-20 12:33:19 quim vwConfig[24241] DEBUG Calling get for qualys_web:db_path
2019-03-20 12:33:19 quim vwConfig[24241] DEBUG Calling getbool for qualys_web:verbose
2019-03-20 12:33:19 quim vulnWhispererBase[24241] INFO Connected to database at ./data/database/report_tracker.db
2019-03-20 12:33:19 quim qualysapi.config[24241] WARNING configs/frameworks_example.ini permissions allows more than user access.
2019-03-20 12:33:19 quim qualysapi.config[24241] DEBUG [('max_retries', '10'), ('hostname', 'qualysapi.qg2.apps.qualys.com'), ('template_id', '126024'), ('enabled', 'true'), ('username', 'exampleuser'), ('password', 'examplepass'), ('write_path', './data/qualys/'), ('db_path', './data/database'), ('verbose', 'true')]
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG proxies = 
None
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG max_retries = 
10
2019-03-20 12:33:19 quim urllib3.util.retry[24241] DEBUG Converted retries value: 10 -> Retry(total=10, connect=None, read=None, redirect=None, status=None)
2019-03-20 12:33:19 quim urllib3.util.retry[24241] DEBUG Converted retries value: 10 -> Retry(total=10, connect=None, read=None, redirect=None, status=None)
2019-03-20 12:33:19 quim qualysapi.util[24241] INFO Finished building connector.
2019-03-20 12:33:19 quim qualysWhisperAPI[24241] INFO Connected to Qualys at qualysapi.qg2.apps.qualys.com
2019-03-20 12:33:19 quim qualysapi.config[24241] WARNING configs/frameworks_example.ini permissions allows more than user access.
2019-03-20 12:33:19 quim qualysapi.config[24241] DEBUG [('max_retries', '10'), ('hostname', 'qualysapi.qg2.apps.qualys.com'), ('template_id', '126024'), ('enabled', 'true'), ('username', 'exampleuser'), ('password', 'examplepass'), ('write_path', './data/qualys/'), ('db_path', './data/database'), ('verbose', 'true')]
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG api_call =
/count/was/wasscan
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG api_version =
None
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG data <type 'lxml.etree._Element'> =
 <Element ServiceRequest at 0x7f69944495a8>
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG http_method =
None
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG concurrent_scans_retries =
0
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG concurrent_scans_retry_delay =
0
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG api_call post strip =
count/was/wasscan
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG Base url =
https://qualysapi.qg2.apps.qualys.com/qps/rest/3.0/
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG headers =
{'X-Requested-With': 'Parag Baxi QualysAPI (python) v5.1.0'}
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG http_method =
post
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG api_call post strip =
count/was/wasscan
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG api_call =
count/was/wasscan
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG Converting lxml.builder.E to string
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG Converted:
<ServiceRequest><filters><Criteria operator="EQUALS" field="status">FINISHED</Criteria></filters></ServiceRequest>
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG url =
https://qualysapi.qg2.apps.qualys.com/qps/rest/3.0/count/was/wasscan
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG data =
<ServiceRequest><filters><Criteria operator="EQUALS" field="status">FINISHED</Criteria></filters></ServiceRequest>
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG headers =
{'Content-type': 'text/xml', 'X-Requested-With': 'Parag Baxi QualysAPI (python) v5.1.0'}
2019-03-20 12:33:19 quim qualysapi.connector[24241] DEBUG POST request.
2019-03-20 12:33:19 quim urllib3.connectionpool[24241] DEBUG Starting new HTTPS connection (1): qualysapi.qg2.apps.qualys.com
2019-03-20 12:33:21 quim urllib3.connectionpool[24241] DEBUG https://qualysapi.qg2.apps.qualys.com:443 "POST /qps/rest/3.0/count/was/wasscan HTTP/1.1" 200 None
2019-03-20 12:33:21 quim qualysapi.connector[24241] DEBUG response headers =
{'Content-Encoding': 'gzip', 'Transfer-Encoding': 'chunked', 'X-Powered-By': 'Qualys:USPOD02:null:null', 'Vary': 'Accept-Encoding', 'Server': 'Apache', 'Date': 'Wed, 20 Mar 2019 11:33:21 GMT', 'Content-Type': 'application/xml'}
2019-03-20 12:33:21 quim qualysapi.connector[24241] DEBUG 'x-ratelimit-remaining'
2019-03-20 12:33:21 quim qualysapi.connector[24241] DEBUG response text =
<?xml version="1.0" encoding="UTF-8"?>
<ServiceResponse xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://qualysapi.qg2.apps.qualys.com/qps/xsd/3.0/was/wasscan.xsd">
  <responseCode>INVALID_CREDENTIALS</responseCode>
  <responseErrorDetails>
    <errorMessage>Authentication failed</errorMessage>
    <errorResolution>Verify both &apos;user&apos; and &apos;password&apos; credentials.</errorResolution>
  </responseErrorDetails>
</ServiceResponse>
2019-03-20 12:33:21 quim qualysWhisperAPI[24241] ERROR Couldn't process all scans: no such child: count
2019-03-20 12:33:21 quim main[24241] ERROR No objects to concatenate
```
As you can see, there is no longer the request for interactive credentials, and at the end it crashes in an error due to having `INVALID CREDENTIALS`, so when trying to process the scans it gives `no such child: count`

Cheers!

**Comment by reactor7171** [comment]
Btw, you are totally right, there is a code duplication at https://github.com/HASecuritySolutions/VulnWhisperer/blob/beta-1.8/vulnwhisp/frameworks/qualys_web.py#L644 regarding https://github.com/HASecuritySolutions/VulnWhisperer/blob/master/vulnwhisp/vulnwhisp.py#L517, this is something that should be fixed.

**Comment by benweber2** [comment]
Hey @qmontal,

Can you confirm my git commands are correct?  don't want to give you misleading information:

`git clone https://github.com/HASecuritySolutions/VulnWhisperer.git`

`git pull https://github.com/HASecuritySolutions/VulnWhisperer.git beta-1.8`

`sudo python setup.py install`


With this, I'm getting a different error:

`
ERROR:qualysWhisperAPI:Already have WAS scan count
INFO:qualysWhisperAPI:Retrieving information for 16280 scans
ERROR:qualysWhisperAPI:Couldn't process all scans: tuple index out of range
ERROR:main:No objects to concatenate
ERROR: No objects to concatenate
`

**Comment by reactor7171** [comment]
Hi @benweber2,

I usually do after the `git clone`, `git checkout beta-1.8`, then it changes directly to that code branch; when you do the clone, you download all of the branches already.

From the error you are getting, I can see we managed to solve the first 2 errors you were having, but as I thought, there are some more issues. 

My main problem is the access to a Qualys WAS instance, as this was developed I guess that by @austin-taylor himself; once we are in this situation I can only try to fix the code by connecting to you and try to fix whatever is breaking there, as once having qualysapi fixed, everything should be working, so probably something changed on the Qualys WAS side.

For this though, it will be better to get in contact through our Slack channel :)

**Comment by reactor7171** [comment]
Will try to fix this for version 1.8 release, although we didn't expect to have this issue and we were just missing the documentation update in order to do the release, so it might get delayed to the next version.
